cmd/snap-exec/main.go

```diff
@@ -199,10 +200,13 @@ func execApp(snapApp, revision, command string, args []string) error {
 	// variables to plugging snap apps, but this is a lot simpler as a
 	// work-around
 	// we currently only handle the CUPS_SERVER environment variable, setting it
-	// to /var/cups/ if that directory exists - it should not exist anywhere
+	// to /var/cups/ if that dir is a bind-mount - it should not be one
 	// except in a strictly confined snap where we setup the bind mount from the
-	// source cups slot snap to the plugging snap
-	if exists, _, _ := osutil.DirExists(dirs.GlobalRootDir + "/var/cups"); exists {
+	// source cups slot snap to the plugging snap.
+	var stVar, stVarCups syscall.Stat_t
+	err1 := syscallStat(dirs.GlobalRootDir+"/var/", &stVar)
+	err2 := syscallStat(dirs.GlobalRootDir+"/var/cups/", &stVarCups)
+	if err1 == nil && err2 == nil && stVar.Dev != stVarCups.Dev {
```

Comment:
This check is not completely correct, because if the same device was mounted both over `/var/` (or `/`) and `/var/cups/` you would not detect the bind-mount and the variable would still be unset. The correct check would be reading `/proc/self/mountinfo`, but maybe there are performance reason not to use that? Or maybe we could even move this logic to snap-confine, which is already parsing mountinfo IIRC.

If you think that this is not going to be a real issue, feel free to disregard my remark (though adding an extra comment on the possible risks would be welcome).

Though, now that I think of it, why don't we just define the CUPS_SERVER variable in all cases? It's not that we are giving up any sensible information, and if the interface is not connected then the socket will just not be exposed...

Reply:
I'm not sure I understand. This check happens in snap-exec so inside the mount namespace of the snap. We know that /var comes from the core{,18,20,22} snap in this case. If there is a bind mount it will come from a different snap so the dev_t is different. This should be reliable unless I miss something.

As for why this is needed - we cannot always set it, because is exactly what broke cups-control. For snaps that use cups-control (or for the cups snap itself) it's important that this is not set or cups will look into the wrong places.

Reply:
Ah, I see. I guess we might still have the issue if a snap using the cups interface specifies a base snap which is the same snap which provides the cups socket, but I guess I'm overthinking it for the sake of it :-)

Reply:
Thanks, great point. I hope/think it's not much of a practical issue  - fwiw, this is only an issue until we have a way to export env vars from an interface, something we want to do in the next cycle.